nflow-tests/src/test/java/io/nflow/tests/ConcurrentEnginesTest.java

```diff
@@ -0,0 +1,124 @@
+public class ConcurrentEnginesTest {
+  private static final int ENGINES = 4;
+  private static final int WORKFLOWS = 100 * ENGINES;
+  private static final String UNIQUE_KEY = UUID.randomUUID().toString();
+  private static final List<NflowServerConfig> servers = new ArrayList<>();
+  private WebClient workflowInstanceResource;
+
+  @ComponentScan(basePackageClasses = DemoWorkflow.class)
+  static class DemoConfiguration {
+    // for component scanning only
+  }
+
+  @BeforeAll
+  public static void start() throws Exception {
+    servers.add(new NflowServerConfig.Builder()
+            .prop("nflow.executor.thread.count",2)
+            .prop("nflow.dispatcher.executor.queue.size", 5)
+            .prop("nflow.executor.group","concur")
+            .prop("nflow.dispatcher.sleep.ms", 5)
+            .prop("nflow.dispatcher.sleep.ms", 2)
+            .springContextClass(DemoConfiguration.class).build());
+    servers.get(0).before();
+    for (int i=1; i<ENGINES; ++i) {
+      NflowServerConfig server = servers.get(0).anotherServer();
+      servers.add(server);
+      server.before();
+    }
+  }
+
+  @Inject
+  public void setWorkflowInstanceResource(@Named("workflowInstance") WebClient client) {
+    String newUri = UriBuilder.fromUri(client.getCurrentURI()).port(servers.get(0).getPort()).build().toString();
+    this.workflowInstanceResource = fromClient(client, true).to(newUri, false);
+  }
+
+  @AfterAll
+  public static void stop() {
+    Collections.reverse(servers);
+    servers.forEach(NflowServerConfig::after);
+  }
+
+  @Test
+  @Order(1)
+  public void insertWorkflows() {
+    CreateWorkflowInstanceRequest req = new CreateWorkflowInstanceRequest();
+    req.type = DEMO_WORKFLOW_TYPE;
+    req.businessKey = UNIQUE_KEY;
+    req.activationTime = now().plusSeconds(10);
+    for (int i=0; i<WORKFLOWS; ++i) {
+      CreateWorkflowInstanceResponse resp = fromClient(workflowInstanceResource, true).put(req, CreateWorkflowInstanceResponse.class);
+      assertThat(resp.id, notNullValue());
+    }
+  }
+
+  @Test
+  @Order(2)
+  public void waitWorkflowsReady() {
+    ListWorkflowInstanceResponse[] wfr = assertTimeoutPreemptively(ofSeconds(60), () -> {
+      ListWorkflowInstanceResponse wf = null;
+      while (true) {
+        sleep(500);
+        ListWorkflowInstanceResponse[] instances = fromClient(workflowInstanceResource, true)
+                .query("type", DEMO_WORKFLOW_TYPE)
+                .query("maxResults", WORKFLOWS + 1)
+                .query("businessKey", UNIQUE_KEY)
+                .query("status", finished)
+                .query("include", "").get(ListWorkflowInstanceResponse[].class);
+        if (instances.length == WORKFLOWS) {
+          return instances;
+        }
+      }
+    });
+    assertThat(wfr.length, is(WORKFLOWS));
+    List<DateTime> startTimes = Stream.of(wfr).map(r -> r.started).sorted().collect(toList());
```

Comment:
startTimes is not used, can this be removed?

Reply:
I was thinking of coming up with something to test with the start times, comparing the max between avg. But could not get any difference with or without deadlocks. What would be a good way to test if the deadlock condition occured?

Reply:
If there's no easy way to test it, just remove it?